cmd/zfs/zfs_project.c

```diff
@@ -227,21 +226,29 @@ zfs_project_handle_dir(const char *name, zfs_project_control_t *zpc,
 	zpc->zpc_ignore_noent = B_TRUE;
 	errno = 0;
 	while (!ret && (ent = readdir(dir)) != NULL) {
+		char *fullname;
+
 		/* skip "." and ".." */
 		if (strcmp(ent->d_name, ".") == 0 ||
 		    strcmp(ent->d_name, "..") == 0)
 			continue;
 
-		if (strlen(ent->d_name) + strlen(name) >=
-		    sizeof (fullname) + 1) {
+		if (strlen(ent->d_name) + strlen(name) >= PATH_MAX + 1) {
 			errno = ENAMETOOLONG;
 			break;
 		}
 
-		sprintf(fullname, "%s/%s", name, ent->d_name);
+		ret = asprintf(&fullname, "%s/%s", name, ent->d_name);
```

Comment:
This should not assign to `ret`. Now we're returning a mix of `0`, `-1`, `-errno`, or the number of characters printed by `asprintf()`.

Reply:
Good catch, fixed.